crates/sui-quorum-driver/src/lib.rs

```diff
@@ -0,0 +1,250 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use arc_swap::ArcSwap;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::time::Instant;
+
+use tokio::sync::mpsc::{self, Receiver, Sender};
+use tokio::sync::Mutex;
+use tokio::task::JoinHandle;
+use tracing::log::{error, warn};
+use tracing::Instrument;
+
+use sui_core::authority_aggregator::AuthorityAggregator;
+use sui_core::authority_client::AuthorityAPI;
+use sui_types::error::{SuiError, SuiResult};
+use sui_types::messages::{
+    CertifiedTransaction, ExecuteTransactionRequest, ExecuteTransactionRequestType,
+    ExecuteTransactionResponse, Transaction, TransactionEffects,
+};
+
+const DEFAULT_PROCESS_TIMEOUT: Duration = Duration::from_secs(30);
+
+enum QuorumTask<A> {
+    ProcessTransaction(Transaction),
+    ProcessCertificate(CertifiedTransaction),
+    UpdateValidators(AuthorityAggregator<A>),
+}
+
+pub struct QuorumDriverHandler<A> {
+    quorum_driver: Arc<QuorumDriver<A>>,
+    _processor_handle: JoinHandle<()>,
+    task_sender: Mutex<Sender<QuorumTask<A>>>,
+    // TODO: Change to CertifiedTransactionEffects eventually.
+    effects_subscriber: Mutex<Receiver<(CertifiedTransaction, TransactionEffects)>>,
+}
+
+struct QuorumDriver<A> {
+    validators: ArcSwap<AuthorityAggregator<A>>,
+}
+
+impl<A> QuorumDriver<A> {
+    pub fn new(validators: AuthorityAggregator<A>) -> Self {
+        Self {
+            validators: ArcSwap::from(Arc::new(validators)),
+        }
+    }
+}
+
+impl<A> QuorumDriverHandler<A>
+where
+    A: AuthorityAPI + Send + Sync + 'static + Clone,
+{
+    pub fn new(validators: AuthorityAggregator<A>) -> Self {
+        let quorum_driver = Arc::new(QuorumDriver::new(validators));
+        let (task_tx, task_rx) = mpsc::channel::<QuorumTask<A>>(5000);
+        let (subscriber_tx, subscriber_rx) = mpsc::channel::<_>(5000);
+        let handle = {
+            let task_tx_copy = task_tx.clone();
+            let quorum_driver_copy = quorum_driver.clone();
+            tokio::task::spawn(async move {
+                Self::task_queue_processor(
+                    quorum_driver_copy,
+                    task_rx,
+                    task_tx_copy,
+                    subscriber_tx,
+                )
+                .await;
+            })
+        };
+        Self {
+            quorum_driver,
+            _processor_handle: handle,
+            task_sender: Mutex::new(task_tx),
+            effects_subscriber: Mutex::new(subscriber_rx),
+        }
+    }
+
+    pub async fn next_effects(&self) -> Option<(CertifiedTransaction, TransactionEffects)> {
```

Comment:
will this only be used by tests? if so can you add `#[cfg(test)]` to make it clear to future readers its not a true public api

Reply:
My plan for this API is that it will eventually be called from the fullnode, so that fullnode can "subscribe" to effects certs coming out of the quorum driver